### PR TITLE
Sync upstream 20201109

### PR DESCRIPTION
**Pull request by vext01**
This syncs us with upstream.

There's one extra merge than usual because I started the sync before the last PR was merged, and rebasing these sync-ups never works (you get wacky conflicts along the upstream history).

I'll point out one part of this change in a moment.

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/sir.rs [hunk 1]
```diff
@@ -269,17 +269,22 @@ impl SirFuncCx<'tcx> {
...
+        let mut values = Vec::new();
+        let mut target_bbs = Vec::new();
+        for (v, t) in targets.iter() {
+            values.push(ykpack::SerU128::new(v));
+            target_bbs.push(t.as_u32());
+        }
         // Create the SwitchInt terminator.
-        let mut targetsnew: Vec<u32> = targets.iter().map(|bb| bb.as_u32()).collect();
-        let otherwise = targetsnew.pop().expect("SwitchInt can't have empty targets?");
+        //let mut targetsnew: Vec<u32> = targets.iter().map(|bb| bb.as_u32()).collect();
```
Comment (vext01) [inline compiler/rustc_codegen_ssa/src/sir.rs]: Oops. I'll kill these comments too.

File: compiler/rustc_codegen_ssa/src/sir.rs [hunk 2]
```diff
@@ -392,11 +397,7 @@ impl SirFuncCx<'tcx> {
     where
         T: TypeFoldable<'tcx> + Copy,
     {
-        if let Some(substs) = self.instance.substs_for_mir_body() {
-            self.tcx.subst_and_normalize_erasing_regions(substs, ty::ParamEnv::reveal_all(), value)
-        } else {
-            self.tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), *value)
-        }
+        self.instance.subst_mir(self.tcx, value)
```
Comment (vext01) [inline compiler/rustc_codegen_ssa/src/sir.rs]: This function was lifted from cranelift and `substs_for_mir_body()` has been removed by upstream.

I looked in the codegen_clif git history to see if @bjorn3 had already handled this, but it doesn't appear so.

What do you reckon @bjorn3? Looks like the right fix? FWIW all of our tests pass.

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/sir.rs]: https://github.com/rust-lang/rust/commit/8a8ee1a3ed249e4077a7ce0b88903b493419e7d7 uses `        self.instance.subst_mir_and_normalize_erasing_regions(self.tcx, ty::ParamEnv::reveal_all(), value)`.

(CI on rust-lang/rust ensures that the cg_clif subtree always compiles. Periodically I pull in the changes to the cg_clif subtree into the main repo of cg_clif and vice versa.)

Comment (vext01) [reply compiler/rustc_codegen_ssa/src/sir.rs]: Ah, thanks.

I'd been looking in https://github.com/bjorn3/rustc_codegen_cranelift -- the wrong place perhaps?

https://github.com/bjorn3/rustc_codegen_cranelift/blob/4700926e54e71002f25c5fbf95dc6d099589837f/src/common.rs#L360-L371

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/sir.rs]: That is the main repo. In https://github.com/rust-lang/rust/pull/77975 it was added as `git subtree` to the rust repo. `git subtree` allows syncing of a directory in a repo with a whole different repo. In this case the `compiler/rustc_codegen_cranelift` directory with the https://github.com/bjorn3/rustc_codegen_cranelift repo.

Comment (vext01) [reply compiler/rustc_codegen_ssa/src/sir.rs]: Thanks for clarification.

**Comment by vext01** [comment]
May I squash?

**Comment by ptersilie** [comment]
Go ahead.

**Comment by vext01** [comment]
Splat.

**Pull request merged**